### PR TITLE
Add a toggle that changes the time frame of jobs aggregation displayed on the dashboard page

### PR DESCRIPTION
**Pull request by zacharyhsia**

### CODE REVIEW

File: servers/dsub/jobs/controllers/utils/jobs_generator.py [hunk 1]
```diff
@@ -95,7 +95,11 @@ def generate_jobs_by_window(provider, project_id, window_min, window_max=None):
     Returns:
         generator: Retrieved jobs
     """
-    create_time_min = window_min - datetime.timedelta(days=_MAX_RUNTIME_DAYS)
+    create_time_min = None
+
+    if window_min is not None:
+        create_time_min = window_min - datetime.timedelta(
+            days=_MAX_RUNTIME_DAYS) if window_min is not None else None
```
Comment (calbach) [inline servers/dsub/jobs/controllers/utils/jobs_generator.py]: window_min is already not None at this point in the code

File: ui/src/app/dashboard/dashboard.component.css [hunk 2]
```diff
@@ -21,4 +21,9 @@ div {
   margin: 1rem;
 }
 
+mat-select {
+/*TODO: width adjust to content*/
+width: 150px;
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.component.css]: spaces

File: ui/src/app/dashboard/dashboard.component.html [hunk 3]
```diff
@@ -1,12 +1,21 @@
 <jm-header [showControls]="false"></jm-header>
-
 <div class="aggregation-container">
-  <jm-total-summary [summary]="aggregationResponse.summary" [statusArray]="statusArray"></jm-total-summary>
+  <span class="title">Total Jobs</span>
+  <mat-select [placeholder]='timeFrameToDescriptionMap.get(selectedTimeFrame)' (selectionChange)="onTimeFrameChange($event.value)">
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.component.html]: nit: selectionChange on next line

File: ui/src/app/dashboard/dashboard.component.ts [hunk 4]
```diff
@@ -11,17 +14,47 @@ import {JobStatus} from "../shared/model/JobStatus";
...
+
+  constructor(private readonly router: Router,
+              private readonly activatedRoute: ActivatedRoute) {}
 
   ngOnInit() {
-    this.aggregationResponse = this.activatedRoute.snapshot.data['aggregations'];
+    this.activatedRoute.queryParams.subscribe(() => this.reloadAggregations());
+
+    // make the default toggle option match to url param
+    if (this.activatedRoute.snapshot.queryParams['timeFrame'] != null) {
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.component.ts]: Shorter and handles the empty string case `if (this.activatedRoute.snapshot.queryParams['timeFrame']) { ...`

File: ui/src/app/dashboard/dashboard.component.ts [hunk 5]
```diff
@@ -11,17 +14,47 @@ import {JobStatus} from "../shared/model/JobStatus";
...
     for (let countEntry of this.aggregationResponse.summary.counts) {
       if (!this.statusArray.includes(countEntry.status)) {
         this.statusArray.push(countEntry.status);
       }
     }
   }
+
+  onTimeFrameChange(newTimeFrame: TimeFrame) {
+    this.selectedTimeFrame = newTimeFrame;
+    const extras = {
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.component.ts]: opt_nit: I would probably inline this below instead

File: ui/src/app/dashboard/dashboard.module.ts [hunk 6]
```diff
@@ -21,7 +21,7 @@ import {ClrIconModule, ClrTooltipModule} from "@clr/angular";
   declarations: [
     DashboardComponent,
     TotalSummaryComponent,
-    GroupedSummaryComponent
+    GroupedSummaryComponent,
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.module.ts]: revert, no need to touch this file in the PR

File: ui/src/app/dashboard/dashboard.resolver.service.ts [hunk 7]
```diff
@@ -1,8 +1,10 @@
+///<reference path="../shared/utils/url-search-params.utils.ts"/>
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.resolver.service.ts]: What is this syntax?

File: ui/src/app/dashboard/dashboard.resolver.service.ts [hunk 8]
```diff
@@ -12,7 +14,7 @@ export class DashboardResolver implements Resolve<AggregationResponse>{
   ) {}
 
   resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<AggregationResponse> {
-    return this.jobManagerService.queryAggregations(TimeFrame.DAYS7,
+    return this.jobManagerService.queryAggregations(stringToTimeFrameMap.get(route.queryParams['timeFrame']),
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.resolver.service.ts]: Doesn't this need to handle the default case as well? What if the query param is not given?

File: servers/dsub/jobs/controllers/utils/jobs_generator.py (hunk 1 above)
Comment (zacharyhsia) [reply servers/dsub/jobs/controllers/utils/jobs_generator.py]: Fixed.

File: ui/src/app/dashboard/dashboard.component.css (hunk 2 above)
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.component.css]: Fixed.

File: ui/src/app/dashboard/dashboard.component.html (hunk 3 above)
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.component.html]: Fixed.

File: ui/src/app/dashboard/dashboard.component.ts (hunk 4 above)
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.component.ts]: Nice!

File: ui/src/app/dashboard/dashboard.component.ts (hunk 5 above)
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.component.ts]: Fixed.

File: ui/src/app/dashboard/dashboard.module.ts (hunk 6 above)
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.module.ts]: Fixed.

File: ui/src/app/dashboard/dashboard.resolver.service.ts (hunk 7 above)
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.resolver.service.ts]: Removed.

File: ui/src/app/dashboard/dashboard.resolver.service.ts (hunk 8 above)
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.resolver.service.ts]: In default case we also encode the timeframe in the URL, we want to redirect user if the timeframe is not in the url. 
Adding a TODO here to reminder the "redirect" thing.

File: ui/src/app/shared/common.ts [hunk 9]
```diff
@@ -36,6 +39,30 @@ export const queryExtensionsDataTypes: Map<string, FieldDataType> = new Map([
...
+  ['HOURS_1', TimeFrame.HOURS1],
+  ['HOURS_8', TimeFrame.HOURS8],
+  ['HOURS_24', TimeFrame.HOURS24],
+  ['DAYS_7', TimeFrame.DAYS7],
+  ['DAYS_30', TimeFrame.DAYS30],
+  ['ALL_TIME', TimeFrame.ALLTIME],
+]);
+
+export const timeFrameToDescriptionMap = new Map<TimeFrame, string> ([
+  [TimeFrame.HOURS1, 'in past 1 hour'],
```
Comment (calbach) [inline ui/src/app/shared/common.ts]: The mocks show "Past 1 hour".

File: ui/src/app/shared/common.ts [hunk 10]
```diff
@@ -36,6 +39,30 @@ export const queryExtensionsDataTypes: Map<string, FieldDataType> = new Map([
...
+  [TimeFrame.HOURS8, 'in past 8 hours'],
+  [TimeFrame.HOURS24, 'in past 24 hours'],
+  [TimeFrame.DAYS7, 'in past 7 days'],
+  [TimeFrame.DAYS30, 'in past 30 days'],
+  [TimeFrame.ALLTIME, 'in past all time'],
```
Comment (calbach) [inline ui/src/app/shared/common.ts]: "past all time" doesn't really make sense, I'd just go with "all time", or possibly "for all time"

File: ui/src/app/shared/common.ts [hunk 11]
```diff
@@ -36,6 +39,30 @@ export const queryExtensionsDataTypes: Map<string, FieldDataType> = new Map([
   ['submission', FieldDataType.Date]
 ]);
 
+export const stringToTimeFrameMap: Map<String, TimeFrame> = new Map([
+  ['HOURS_1', TimeFrame.HOURS1],
```
Comment (calbach) [inline ui/src/app/shared/common.ts]: What is the string here? Why would it be different from the enum value which is already mapped for us from swagger codegen?

File: ui/src/app/shared/utils/url-search-params.utils.ts [hunk 12]
```diff
@@ -111,6 +112,31 @@ import {FieldDataType, queryDataTypes, queryExtensionsDataTypes} from "../common
     return queryRequest;
   }
 
+  /** Get the start time by the given time frame on the dashboard page */
+  public static getStartTimeByTimeFrame(timeFrame: TimeFrame): string {
+    let now = new Date;
```
Comment (calbach) [inline ui/src/app/shared/utils/url-search-params.utils.ts]: const

File: ui/src/app/dashboard/total-summary/total-summary.component.ts [hunk 13]
```diff
@@ -1,39 +1,63 @@
...
+    this.statusCountsMap = newStatusCountsMap;
   }
 
   getUrlParams(status: JobStatus) {
-    let query = URLSearchParamsUtils.unpackURLSearchParams(this.activatedRoute.snapshot.queryParams['q']);
-    query.statuses = [status];
-    return {q: URLSearchParamsUtils.encodeURLSearchParams(query)};
+    const query = URLSearchParamsUtils.unpackURLSearchParams(this.activatedRoute.snapshot.queryParams['q']);
+
+    let map = new Map<string, string[]>();
```
[the remark this answers is not in the document]
Comment (zacharyhsia) [reply ui/src/app/dashboard/total-summary/total-summary.component.ts]: Nice! Also, FYI, I tried to also use this in grouped-summary but labels are not strong-typed so stay with map in that case.

[the remark this answers is not in the document]
Comment (zacharyhsia) [reply ui/src/app/dashboard/total-summary/total-summary.component.ts]: Nice! Also, FYI, I tried to also use this in grouped-summary but labels are not strong-typed so stay with map in that case.

File: ui/src/app/shared/common.ts (hunk 11 above)
Comment (zacharyhsia) [reply ui/src/app/shared/common.ts]: How to use that to map a string to TimeFrame object?

File: ui/src/app/shared/common.ts (hunk 9 above)
Comment (zacharyhsia) [reply ui/src/app/shared/common.ts]: Fixed.

File: ui/src/app/shared/common.ts (hunk 10 above)
Comment (zacharyhsia) [reply ui/src/app/shared/common.ts]: Using "All time".

File: ui/src/app/shared/common.ts (hunk 11 above)
Comment (bfcrampton) [reply ui/src/app/shared/common.ts]: In [`dashboard.resolver.service.ts`](https://github.com/DataBiosphere/job-manager/pull/453/files#diff-573e628c01e7c1d012851d65411063b4R18):

```
timeFrame = (<any>TimeFrame)[route.queryParams['timeFrame']]
```

File: ui/src/app/shared/common.ts [hunk 14]
```diff
@@ -36,6 +39,30 @@ export const queryExtensionsDataTypes: Map<string, FieldDataType> = new Map([
...
+  ['HOURS_8', TimeFrame.HOURS8],
+  ['HOURS_24', TimeFrame.HOURS24],
+  ['DAYS_7', TimeFrame.DAYS7],
+  ['DAYS_30', TimeFrame.DAYS30],
+  ['ALL_TIME', TimeFrame.ALLTIME],
+]);
+
+export const timeFrameToDescriptionMap = new Map<TimeFrame, string> ([
+  [TimeFrame.HOURS1, 'Past 1 hour'],
```
Comment (bfcrampton) [inline ui/src/app/shared/common.ts]: Nit: Prefer "Past Hour"

File: ui/src/app/app-routing.module.ts [hunk 15]
```diff
@@ -39,8 +39,9 @@ const routes: Routes = [
   {
     path: 'dashboard',
     component: DashboardComponent,
-    //TODO: (zach) dashboard need a query param of project id before it can be activated
+    //TODO: (zach) dashboard need query params of projectId also timeFrame before it can be activated
```
Comment (bfcrampton) [inline ui/src/app/app-routing.module.ts]: Put the action item in the todo, what needs to be fixed? Is it that if you route to /dashboard without project you aren't properly redirected to the projects page?

File: ui/src/app/dashboard/dashboard.component.css [hunk 16]
```diff
@@ -21,4 +21,9 @@ div {
   margin: 1rem;
 }
 
+mat-select {
+/*TODO: width adjust to content*/
```
[the remark this answers is not in the document]
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.component.css]: Cool!

File: ui/src/app/shared/common.ts (hunk 14 above)
Comment (zacharyhsia) [reply ui/src/app/shared/common.ts]: Fixed.

File: ui/src/app/shared/common.ts (hunk 11 above)
Comment (zacharyhsia) [reply ui/src/app/shared/common.ts]: When we encode the timeFrame in the url, the server find this parameter to be an unicode string, which dose not match to enum type in python. Should we do a conversion in javascript or python?

File: ui/src/app/dashboard/dashboard.component.ts [hunk 17]
```diff
@@ -11,17 +14,46 @@ import {JobStatus} from "../shared/model/JobStatus";
 export class DashboardComponent implements OnInit {
   aggregationResponse: AggregationResponse;
   statusArray: Array<JobStatus> = [JobStatus.Succeeded, JobStatus.Aborted, JobStatus.Running, JobStatus.Failed];
+  timeFrameToDescriptionMap = timeFrameToDescriptionMap;
```
[the remark this answers is not in the document]
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.component.ts]: Accepted.

File: ui/src/app/app-routing.module.ts (hunk 15 above)
Comment (zacharyhsia) [reply ui/src/app/app-routing.module.ts]: Yes. If the projectId is missing, it shows 400 error. Adding this action to the TODO.

File: ui/src/app/shared/common.ts (hunk 11 above)
Comment (bfcrampton) [reply ui/src/app/shared/common.ts]: Ah sorry what I posted was wrong, all you actually have to do is a type conversion in the javascript:

```
timeFrame = <TimeFrame>route.queryParams['timeFrame']
```

Comment (zacharyhsia) [reply ui/src/app/shared/common.ts]: It works! Thanks!

File: ui/src/app/shared/utils/url-search-params.utils.ts [hunk 18]
```diff
@@ -111,6 +112,31 @@ import {FieldDataType, queryDataTypes, queryExtensionsDataTypes} from "../common
...
 
+  /** Get the start time by the given time frame on the dashboard page */
+  public static getStartTimeByTimeFrame(timeFrame: TimeFrame): Date {
+    const now = new Date;
+    switch (timeFrame) {
+      case TimeFrame.HOURS1:
+        now.setHours(now.getHours() - 1);
+        break;
+      case TimeFrame.HOURS8:
+        now.setHours(now.getHours() - 24);
```
Comment (calbach) [inline ui/src/app/shared/utils/url-search-params.utils.ts]: 8

Comment (calbach) [inline ui/src/app/shared/utils/url-search-params.utils.ts]: I don't think this is going to work if hours <= 8

File: ui/src/app/app-routing.module.ts [hunk 19]
```diff
@@ -39,8 +39,10 @@ const routes: Routes = [
   {
     path: 'dashboard',
     component: DashboardComponent,
-    //TODO: (zach) dashboard need a query param of project id before it can be activated
+    //TODO: (zach) if the projectId param is missing, it gives a 400 error which is not desired.
```
Comment (calbach) [inline ui/src/app/app-routing.module.ts]: Could you file an issue for this as well?

File: ui/src/app/dashboard/dashboard.component.html [hunk 20]
```diff
@@ -1,12 +1,22 @@
...
+  <span class="title">Total Jobs</span>
+  <mat-select [placeholder]='tfDescMap.get(selectedTimeFrame)'
+              (selectionChange)="onTimeFrameChange($event.value)">
+    <mat-option *ngFor="let timeFrame of timeFrames" [value]="timeFrame">
+      {{tfDescMap.get(timeFrame)}}
+    </mat-option>
+  </mat-select>
+
+  <jm-total-summary [summary]="aggregationResponse.summary"
+                    [statusArray]="statusArray"
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.component.html]: This reminds me of the statusArray issue (where the card renders poorly for >4 statuses). Can you file an issue for this as well?

File: ui/src/app/shared/utils/url-search-params.utils.ts (hunk 18 above)
Comment (zacharyhsia) [reply ui/src/app/shared/utils/url-search-params.utils.ts]: It dose work. I tested it: now is UTC 8/8 22:00, I subtract by 23 and it gives 8/7 23:00.

File: ui/src/app/dashboard/dashboard.component.html (hunk 20 above)
Comment (zacharyhsia) [reply ui/src/app/dashboard/dashboard.component.html]: Done.

File: ui/src/app/app-routing.module.ts (hunk 19 above)
Comment (zacharyhsia) [reply ui/src/app/app-routing.module.ts]: Sure, done.

File: ui/src/app/shared/utils/url-search-params.utils.ts (hunk 18 above)
Comment (zacharyhsia) [reply ui/src/app/shared/utils/url-search-params.utils.ts]: Fixed.

**Pull request merged**